src/uct/ib/rdmacm/rdmacm_iface.c

```diff
@@ -70,6 +71,230 @@ static uct_iface_ops_t uct_rdmacm_iface_ops = {
+    /* set a new ep (take one from the pending eps queue) */
+    while (!ucs_list_is_empty(&iface->pending_eps_list) && (!set_new_ep)) {
+        iface->ep = ucs_list_extract_head(&iface->pending_eps_list,
+                                          uct_rdmacm_ep_t, list_elem);
+        iface->ep->is_on_pending = 0;
+
+        if (rdma_resolve_addr(iface->cm_id, NULL, iface->ep->remote_addr,
+                              UCS_MSEC_PER_SEC * iface->config.addr_resolve_timeout)) {
+            ucs_error("rdma_resolve_addr(addr=%s) failed: %m",
+                       ucs_sockaddr_str(iface->ep->remote_addr, ip_str, ip_len));
+            uct_rdmacm_ep_set_failed(&iface->super.super, &iface->ep->super.super,
+                                     UCS_ERR_IO_ERROR);
+        } else {
+            set_new_ep = 1;
+        }
+    }
+
+    if (!set_new_ep) {
+        iface->ep = NULL;
+    }
+
+    UCS_ASYNC_UNBLOCK(iface->super.worker->async);
+}
+
+static void uct_rdmacm_client_handle_failure(uct_rdmacm_iface_t *iface, ucs_status_t status)
+{
+    ucs_assert((!iface->is_server));
+    uct_rdmacm_ep_set_failed(&iface->super.super, &iface->ep->super.super, status);
+    uct_rdmacm_iface_client_start_next_ep(iface);
+}
+
+static void uct_rdmacm_iface_process_conn_req(uct_rdmacm_iface_t *iface,
+                                             struct rdma_cm_event *event,
+                                             struct sockaddr *remote_addr)
+{
+    uct_rdmacm_priv_data_hdr_t *hdr, cb_hdr;
+    size_t max_conn_priv;
+    ssize_t server_data_len;
+    struct rdma_conn_param conn_param;
+    ucs_status_t status;
+    size_t ip_len  = ucs_max(INET_ADDRSTRLEN, INET6_ADDRSTRLEN);
+    char *ip_str   = ucs_alloca(ip_len);
+
+    status = uct_base_iface_get_max_conn_priv(&iface->super.super, &max_conn_priv);
+    if (status != UCS_OK) {
+        ucs_error("Server failed to get max_conn_priv");
+        rdma_reject(event->id, NULL, 0);
+        return;
+    }
+
+    hdr = (uct_rdmacm_priv_data_hdr_t*) event->param.ud.private_data;
+
+    memset(&conn_param, 0, sizeof(conn_param));
+    conn_param.private_data = ucs_alloca(max_conn_priv + sizeof(uct_rdmacm_priv_data_hdr_t));
+
+    /* TODO check the iface's cb_flags to determine when to invoke this callback.
+     * currently only UCT_CB_FLAG_ASYNC is supported so the cb is invoked from here */
+    server_data_len = iface->conn_request_cb(iface->conn_request_arg,
+                                             event->param.ud.private_data +
+                                             sizeof(uct_rdmacm_priv_data_hdr_t), /* private data */
+                                             hdr->length,                        /* length */
+                                             (void*)(conn_param.private_data +
+                                             sizeof(uct_rdmacm_priv_data_hdr_t)));
+    if (server_data_len < 0) {
+        ucs_error("Server failed to fill private data");
```

Comment:
1. pls start log messages with small letters
1. pls add more details (e.g that it comes from rdmacm)